src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -41,15 +41,108 @@ private GCHandle(object? value, GCHandleType type)
                 throw new ArgumentException(SR.ArgumentException_NotIsomorphic, nameof(value));
             }
 
-            IntPtr handle = InternalAlloc(value, type);
+            _handle = type == GCHandleType.Pinned ?
+                AllocPinnedWithPooling(value) :
+                InternalAlloc(value, type);
```

Comment:
This will impact profiler APIs and event tracing. Profiler APIs are not going to see all handle allocations and free as before. E.g. if the diagnostic tools has feature to detect handle leaks, the stacktraces associated with the leaking handles are going to be bogus.

Reply:
handle creation/free will be the same as before - you still get the callback and the stress log msg when a handle is allocated from the handle table and freed to the handle table. and you can already change the Target of a handle today so this doesn't introduce a new scenario. and there's an event that gets fired when this happens.

Reply:
I am talking about these profiler callbacks https://github.com/dotnet/runtime/blob/cbc501ca196371572c38f8d12a66969864d99c08/src/coreclr/inc/corprof.idl#L2311-L2321 . I believe that these callbacks were specifically designed to allow profilers to capture callstacks for when the GC handle is created and freed and detect leaking handles.

The cache of GC handles breaks the diagnostic value of these callbacks. The profiler will only get the callback when the handle is allocated for the first time from the GC handle table. I won't get callbacks when the handle is reused and so it won't be able to capture the user code callstack that allocated the handle.

Reply:
yeah, I knew that's what you meant - those are the only profiling API callbacks for handle creation/free. 

today you can create a bunch of handles and just change their targets, including setting the targets to null which is exactly what happens here. I don't see this as a blocking factor for doing perf improvement. if this is really needed, we could add another callback when target is changed.